### DISCUSSION TITLE
Gradle manager sets incorrect depType for test dependencies

### DISCUSSION THREAD

**Discussion by AndreasPetersen** [renovatebot/renovate#27777] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub

### Please tell us more about your question or problem

The Gradle manager sets `depType=dependencies` for test dependencies, e.g. both

```
dependencies {
    implementation 'org.apache.commons:commons-lang3:3.9'
    testImplementation 'org.wiremock:wiremock:3.3.0'
}
```

get their `depType` set to `dependencies`. This makes it impossible to distinguish between dependency types.

I have provided a reproducer [here](https://github.com/AndreasPetersen/renovate-26645). I added the `depType` to the `prBodyNotes`, so it is easy to see the selected `depType`.

- [PR for a Gradle plugin](https://github.com/AndreasPetersen/renovate-26645/pull/8) sets `depType=plugin` as expected
- [PR for an implementation dependency](https://github.com/AndreasPetersen/renovate-26645/pull/7) sets `depType=dependencies` as expected
- [PR for a testImplementation dependency](https://github.com/AndreasPetersen/renovate-26645/pull/5) sets `depType=dependencies` not as expected

I'm using Renovate version 37.231.0.

I can see that `depType` can under some circumstances be set to `devDependencies`, but not in the expected case as described above.

https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gradle/extract.ts#L180

I think it would make sense to set the `devType` to the Gradle dependency type, e.g. `implementation`, `testImplementation`, `compileOnly`, `api`, etc.

### Logs (if relevant)

Run the reproducer, if required.

**Comment by Churro** [comment, marked as answer]
This is currently not supported by the `gradle` manager. In contrast to maven, gradle allows plugins to set arbitrary configurations (= dep types). The ones shown in the example snippet only exist when the Java plugin is used.

There are a few things that make this challenging:
* It would be possible to match simple cases like the ones shown using a predefined list of known dep types but very often the dependency name doesn't occur next to the gradle configuration type, and then an intricate source-sink analysis is necessary. 
* Since the gradle manager also does heuristic matching, it may not be feasible to figure out the configuration, or the dependency declaration could be stored in a variable, which is then used multiple times with different configurations.
* renovate currently parses gradle version catalogs but doesn't even look where (and if) a dependency is used. Taking all deps extracted from version catalogs and then trying to pinpoint them across all gradle files is error-prone and negatively impacts runtime, since the parser cannot know, in which file and context a variable is used.

**Comment by AndreasPetersen** [comment]
Thanks for the detailed reply @Churro . I thought it would be simple to take the declared dependency type next to the dependency name, but as you describe it's not that simple.

Perhaps the [Gradle manager documentation](https://docs.renovatebot.com/modules/manager/gradle/#open-items) should state this limitation?

**Comment by AndreasPetersen** [comment]
Would it make sense to cover some cases, and then set an unknown or empty `depType` for more advanced cases? The documentation could then state the supported and unsupported cases.

**Reply by Churro** [reply]
Not sure about this. Even if common configuration types like `implementation` or `testImplementation` are attempted to be matched, it may work in some cases where the dependency name occurs next to the configuration type but fail in other places even in the same project. I guess as a user who is not familiar with how the `gradle` manager works would assume this to work everywhere and not sporadically if support is mentioned in the documentation.
